src/wallet/wallet.cpp

```diff
@@ -1417,6 +1418,57 @@ bool CWallet::IsHDEnabled() const
     return !hdChain.masterKeyID.IsNull();
 }
 
+void CWallet::EventuallyRescanAfterKeypoolTopUp() {
+    if (fSyncPausedUntilKeypoolExt) {
+
+        // for now, enable block requests and tip updates via the states
+        // this will only be sufficient as long as only a single wallet adjusts these stats
+        // otherwise the net-/validation-logic needs to poll the state over a signal (or similar)
+        ::setBlockRequestsPaused(false);
+        ::setTipUpdatesPaused(false);
+
+        // disabled the per-wallet pause
+        fSyncPausedUntilKeypoolExt = false;
+
+        const CChainParams& chainparams = Params();
+        CValidationState state;
+        if (!ActivateBestChain(state, chainparams)) {
+            LogPrintf("Failed to connect best block");
+            StartShutdown();
+        }
+
+        CBlockIndex *pindexRescan = chainActive.Genesis();
+        CWalletDB walletdb(*dbw);
+        CBlockLocator locator;
+        if (walletdb.ReadBestBlock(locator))
+            pindexRescan = FindForkInGlobalIndex(chainActive, locator);
+
+        if (chainActive.Tip() && chainActive.Tip() != pindexRescan)
+        {
+            //We can't rescan beyond non-pruned blocks, stop and throw an error
+            //this might happen if a user uses a old wallet within a pruned node
+            // or if he ran -disablewallet for a longer time, then decided to re-enable
+            if (fPruneMode)
+            {
+                CBlockIndex *block = chainActive.Tip();
+                while (block && block->pprev && (block->pprev->nStatus & BLOCK_HAVE_DATA) && block->pprev->nTx > 0 && pindexRescan != block)
+                    block = block->pprev;
+
+                if (pindexRescan != block) {
+                    const static std::string pruneError = "Prune: last wallet synchronisation goes beyond pruned data. You need to -reindex (download the whole blockchain again in case of pruned node)";
+                    uiInterface.ThreadSafeMessageBox(pruneError, "", CClientUIInterface::MSG_ERROR);
+                    StartShutdown();
```

Comment:
I guess it already tries to avoid this situation by disabling the block downloads if the wallet is encrypted and the node runes in prune-mode.
But there is no guarantee that we won't prune bellow the wallets best block,.. this why this shutdown/protection is here.